### DISCUSSION TITLE
New view transition events

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#740] [Proposal]
### Body

# Summary

Add more events in the view transition router to give users greater control.

# Background & Motivation

The ViewTransitions router currently omits 2 events, but we have gotten requests for more in the past. This proposal is to discuss new potential events.

# Goals

- Add new events for use-cases not covered by current events.
- Allow certain aspects of the view transitions router to be customized through events.

# Events

> Note that this table will be updated based discussion below.

Name                | When called                      | Use-case
--------------------|----------------------------------|---------------------------------------------------------------------------------------------------
`astro:before-swap` | Immediately before a swap occurs | Allows the handler to cancel the event, or to customize swap (with their own swap method perhaps).

**Comment by ematipico** [comment]
Are the events meant to be blocking or not blocking?

**Reply by martrapp** [reply]
If we want to support things like preloading images from the target page, which are needed in the transition, or replacing parts of the page with partials to define the target DOM, then we need a mechanism for the user code to make asynchronous calls. But we also need to wait for the results of the user code before proceeding with the router code. The main points that need to be synchronized can be seen in the diagram below: 
- First red arrow: We have done all the time-consuming preparations and are now ready to freeze the current view.
- Second red arrow: My DOM is exactly as it should be displayed at the end of the view transition. Start the transition.

**Comment by martrapp** [comment]
The following diagram shows a simplified representation of the status quo.  When `navigate()` is called, several preparatory steps are performed before the actual **view transition**, such as loading the target page.

The upper red arrow is when the view transition API takes the "screenshot" for the `before` image, and the lower red arrow marks the point when the DOM got updated and defines the `after` state for the view transition. 

The view transition itself should not be delayed by loading the DOM or similar, as the view would appear frozen during this time. There should also be no new layout or rendering of the current page between these two arrows, as this would be seen as flickering.  

The green blocks are where the current events are dispatched.

```mermaid
graph TD
  subgraph rout[router.ts]
    subgraph API
      direction LR
      nav("navigate()")
      subgraph Events
        eas("astro:after-swap")
        eal("astro:load")
      end
    end

    subgraph flow[Processing Flow]
      osp{"same<br>page?"}
      ff("Load newDocument from URL")
      hco("Postprocess newDocument")    
      ssp("Save scroll position")
      subgraph VTR["<b>View Transition</b>"]
        psl("Preload style links")
        subgraph Swap
          swb("Merge &lt;html>-attributes,<br>&lt;head> and &lt;body><br>of current and newDOM")
          mpe("Copy over persisted elements")
          uhsp2("Move to Location")
          emas("Dispatch astro:after-swap")
        end
      end
      uhsp["Move to Location"]
      flowend["end"]
      RS("Run Scripts")
      empl("Dispatch astro:page-load")
      AN("Announce")
    end
  end
  nav-->osp--"no"-->ff-->hco-->ssp-->psl-->swb-->mpe-->uhsp2-->emas-->RS-->empl-->AN-->flowend
  osp--"yes"-->uhsp------>flowend
   style emas fill:#060,stroke:#333,stroke-width:4px
   style empl fill:#060,stroke:#333,stroke-width:4px
   linkStyle 4,9 stroke:#f00
```

**Reply by martrapp** [reply]
For historical reasons we have a short cut for navigation on the same page, i.e. a transition that does not need to load a new DOM. It just updates the browser history and scrolls to the target scroll position. It does not fire any events.

When we want to support scenarios where users can give alternative implementation to creating a new DOM, swapping it, or even just modifying the existing DOM, it would be more natural to process all navigation the same without the distinction between intra- and extra-page targets.

**Comment by martrapp** [comment]
The design principles for View Transitions are an easy-to-use solution with only a few lines of client-side code.

While this works well in most cases, there are many innovative ideas on how to extend the current solution. While some of these are suitable for most users, others could be considered niche requirements that are beyond the scope and maintainability of the core implementation.

I would like to add some scenarios from users about what they expect from view transitions. We could use those to evaluate the new API. Some of them might become part of the core implementation. But if the new API supports it, some requirements that we don't see in the core could also be realized as add-on and community solutions. 

- Swap that does not lose iframe status. (#717)
- Control over participation in transitions for off-screen elements. (#714)
- Finer control over how elements transition in and out. (#628)
- Improved prefetching with ViewTransitions. (#661)
- Finer control over transition behavior. (#624)
- Progress bar for HTML loading. [#feedback-ideas](https://discord.com/channels/830184174198718474/872579324446928896/1159110175798607942)
- Do nifty things with the [ViewTransition promises](https://developer.chrome.com/docs/web-platform/view-transitions/#animating-with-javascript)

**Reply by matthewp** [reply]
@bluwy is working on moving prefetch to core. As part of that he'll probably also be taking a look into how prefetch is used in VT so some of those configuration ideas might be happening soon.

I think all of those things are unrelated to events though.

**Comment by martrapp** [comment]
For the API extension I would like to see something like the following:

```mermaid
graph TD
  subgraph rout[router.ts]
    subgraph API
      direction TB
      nav("navigate()")
      subgraph Events
      epr("astro:transition-prepare")
      ebs("astro:before-swap")
      eas("astro:after-swap")
      eal("astro:load")
      end
    end

    subgraph flow[Processing Flow]
      emtp("Fire astro:transition-prepare")
      subgraph newDOM["Prepare Transition"]
        ff("Load newDocument from URL")
        hco("Postprocess newDocument")
      end
      ssp("Save scroll position")
      psl("Preload style links")
      subgraph VTR["<b>View Transition</b>"]
        embs("Fire astro:before-swap")
        subgraph Swap
          direction TB
          swb("Merge &lt;html>-attributes,<br>&lt;head> and &lt;body><br>of current and newDOM")
          mpe("Move persisted elements")
          uhsp2("Move to Location")
        end
        emas("Fire astro:after-swap")
      end
      flowend["end"]
      RS("Run Scripts")
      empl("Fire astro:page-load")
      AN("Announce")
    end
  end
  nav-->emtp-->ff-->hco-->ssp-->psl-->embs-->swb-->mpe-->uhsp2-->emas-->RS-->empl-->AN-->flowend
   style emtp fill:#cc0,stroke:#333,stroke-width:4px,color:#000
   style embs fill:#cc0,stroke:#333,stroke-width:4px,color:#000
   style emas fill:#060,stroke:#333,stroke-width:4px,color:#fff
   style empl fill:#060,stroke:#333,stroke-width:4px,color:#fff
   linkStyle 5,10 stroke:#f00
```

**Reply by martrapp** [reply]
The yellow blocks are new interaction points. 
- The `preparation` is called immediately after the call to navigate(). Its job is to do all the preparation necessary before the view transition (browser implementation or our simulation) can begin. It can bypass the standard preparation steps (in the prepare transition box) that load a DOM from a file (e.g. by calling `event.preventDefault()`) and/or it can call/reuse functions provided by the router for some of the standard handling. It may also decide against view transitions and request a full page load.
    - The event would provide the URL of the origin URL and the destination URL of the transition. This allows the user code to decide to do something different for each pair if needed.
    - The event would provide an info object that was previously passed to navigate() to control the transition, see https://developer.mozilla.org/en-US/docs/Web/API/NavigateEvent/info
    - The event would provide a navigation type, see https://developer.mozilla.org/en-US/docs/Web/API/NavigateEvent/navigationType
- `before-swap` is called right after the browsers view transition implementation took its "screenshot" (or in case of simulated transition right after the simulation starts). It can bypass the standard swap steps (in the Swap box) that swap the DOM (e.g. by calling `event.preventDefault()`) and/or it can call/reuse functions provided by the router for some of the standard handling (e.g. swapHead()).
Same event attributes as above + a way to access data provided by the first event.

**Reply by martrapp** [reply]
There might also be a need to distinguish between "at the end of the preparation right before the screenshot" and "at the beginning of the view transition right after the screen shot". In that case the could be another event after `preload style links`.

**Reply by matthewp** [reply]
Sounds great! So as far as use-cases, `before-swap` seems clear to me, it's so that you can define your own swap logic. One note here is that this event will need some of the event about the navigation taking place (like the URL navigating to), as it might want to only have its own swap behavior on certain URLs.

What is a use-case for `transition-prepare`?

**Reply by martrapp** [reply]
What can you do in Prepare?
- I could decide against the default loader/parser and have an alternative that emits events for a progress bar.
- I could temporarily disable animations for elements outside the view port (as an optimization or to prevent annoying fly-throughs). 
- I could examine the origin and destination URLs and based on that dynamically change the way elements are animated (effect, direction, speed, ...) 
- I could look into the newDOM and preload the images that are the target of morph transitions (looks really ugly on first visit with slow network if only loaded during transition)
- I could derive the URL of a partial from the target URL and load that instead of the page. Then I insert it into the existing DOM in the before-swap handler.
I'm sure that event can be a lot of fun.

**Comment by martrapp** [comment]
This is my proposal for the new events:

Names, as always, are up for discussion and I welcome the support of the experts here. Of course, this also applies to all other aspects of the proposal!

There are two new events, `TransitionPrepareEvent` and `TransitionBeforeSwapEvent`.
The events are inspired by the [NavigateEvent of the Navigation API](https://developer.mozilla.org/en-US/docs/Web/API/NavigateEvent), at least as far as that makes sense as long as we don't have broad support for the Navigation API in common browsers.

Both events inherit from a common `TransitionEvent`
```
TransitionEvent extends Event
  readonly origin: URL;  // this is the URL of the current page (important for history navigation as window.location already shows the destination URL)
           destination: URL; // this is the URL of the next page. This will be shown in the browser's addressbar after navigation. Writeable to reflect redirections. 
  readonly navigationType: NavigationType; // push, replace, or traverse
  readonly info: any; // an object that can be passed in the options of navigate()
  readonly canIntercept = true; // all our navigations can be intercepted
  readonly intercept: (handler: () => Promise<void>) => void; // the event handler can register functions that should be executed after (or instead of) the loading of the newDocument but before the view transition starts
           newDocument: Document;  // the current state of the new Document (initially an empty Document)
```
There are also the typical members inherited from Event, with `preventDefault()` / `defaultPrevented` being the most important.

`TransitionPrepareEvent` has additional properties:
```
TransitionPrepareEvent extends TransitionEvent
           direction: string; // 'forward' | 'back', might also be altered
```
`TransitionBeforeSwapEvent` also defines an additional property
```
TransitionBeforeSwapEvent extends TransitionEvent
  readonly viewTransition: ViewTransition; // the ViewTransition object that was returned by startViewTransition(). It holds three promises that inform about the progress of the transition and a function that allows to opt out of the animation during the transition, see [https://developer.mozilla.org/en-US/docs/Web/API/ViewTransition](https://developer.mozilla.org/en-US/docs/Web/API/ViewTransition)
```

**Comment by martrapp** [comment]
The intercept calls allow additional steps to be added to the processing flow. Basically you can get:

```mermaid
graph TD
    subgraph flow[Processing Flow]
      emtp("Fire astro:transition-prepare")
      subgraph newDOM["Prepare Transition"]
        ff("Load newDocument from URL")
        hco("Postprocess newDocument")
      end
      ssp("Save scroll position")
      psl("Preload style links")
      subgraph VTR["<b>View Transition</b>"]
        embs("Fire astro:before-swap")
        subgraph Swap
          direction TB
          swb("Merge &lt;html>-attributes,<br>&lt;head> and &lt;body><br>of current and newDOM")
          mpe("Move persisted elements")
          uhsp2("Move to Location")
        end
        emas("Fire astro:after-swap")
      end
    end
   emtp-->ff-->hco-->pi1("prepare intercept 1")-->pi2("prepare intercept 2")-->pi3("...")-->ssp-->psl-->embs-->swb-->mpe-->uhsp2-->si1("swap intercept 1")-->si2("swap intercept 2")-->si3("...")-->emas
   style emtp fill:#cc0,stroke:#333,stroke-width:4px,color:#000
   style embs fill:#cc0,stroke:#333,stroke-width:4px,color:#000
   style emas fill:#060,stroke:#333,stroke-width:4px,color:#fff
   style newDOM fill:#606,stroke:#333,stroke-width:4px,color:#fff
   style Swap fill:#606,stroke:#333,stroke-width:4px,color:#fff
   linkStyle 7 stroke:#f00
```
The purple blocks are the default implementations for prepare and swap. Calling `preventDefault` removes them from the processing flow. 

If you want to replace the default document loading with your own DOM manipulation/creation function, you would call `preventDefault` in the handler for `TransitionPrepareEvent` and provide your document by calling `e.provideNewDocument`. If you need to do async stuff to provide your document do not create it in the event handler but call `e.intercept` to add a flow step.   

If you want to replace the default swap function with your own, call `preventDefault` in the handler for `TransitionBeforeSwapEvent` and call `e.intercept` to add a flow step that merges `e.newDocument` into `document`.

The the interception steps are executed in order. The next step is only executed once the Promise of its predecessor fulfills.

**Reply by matthewp** [reply]
So `e.intercept()` adds *new* steps, it doesn't replace the internal step?

**Reply by matthewp** [reply]
Ok, I think I understand, if you want to replace the internal step you use `preventDefault`.

Is this modeled after the navigation API intercept? https://developer.mozilla.org/en-US/docs/Web/API/NavigateEvent/intercept

Does that one work the way you're describing?

**Reply by martrapp** [reply]
Yes, I have tried to stick very closely to the navigation API. Once it is widely adopted, I would like to get rid of the history API. I have no insight into the implementation details of `intercept` in the Navigation API, but what I can gather from the examples is the same intent: asynchronous code triggered by the event listener and executed during navigation. However, the semantics of `preventDefault` is different. And we may also need an option for `intercept()` that allows insertion before the default step. ~~The official documentation and specifications on this topic are rather scarce at the moment.~~

Spec: https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-api
Original (appHistory) proposal: https://github.com/WICG/navigation-api
Early Polyfil for the original proposal: https://github.com/frehner/appHistory

**Reply by matthewp** [reply]
Another option (not sure if better), would be that `intercept()` does cause `preventDefault`. We could then export the default step so the user could control ordering that way. Might simplify things?

```js
import { prepareTransition } from "astro:transitions/client";

  document.addEventListener('astro:transition-prepare', (event: Event) => {
    event.intercept(async () => {
      await preloadImages(event.getNewDocument());
      
      // Do the default after
      await prepareTransition(event);
    });
  });
```

**Reply by martrapp** [reply]
😆 Haha, yes, that was my first thought on this topic! Then I was too lazy to think of a good structure to offer the existing functionality. I'll take another look at it. In the end, it's actually a lot easier to explain.

**Comment by martrapp** [comment]
Here is a TypeScript example what it would look like to add a flow step in the preparation phase that preloads the pictures in the new document before the view transition starts (see discord #support thread "First time view transitions between small and big images")

```
import { TRANSITION_PREPARE, isTransitionPrepareEvent } from "astro:transitions/client";

  document.addEventListener(TRANSITION_PREPARE, (event: Event) => {
    if (isTransitionPrepareEvent(event)) {
      event.intercept(async () => {
        preloadImages(event.getNewDocument());
      });
    }
  });
```
The `if (isTransitionPrepareEvent(event))` guard is only for type safety in TypeScript. Pure JavaScript would go without that line.

**Comment by martrapp** [comment]
**Changed in the Event definition above:** 
- Writable properties for `destination` & `direction`
- `origin` and `destination` changed from `string` to `URL`
 
**To be discussed:** Normally event properties are read-only to separate state from signaling. We could use setters/getters as in `preventDefault()` and `defaultPrevented`, but that would just bloat the object and not really improve the state separation.

**Comment by martrapp** [comment]
With regard to further compatibility with the navigation API:

Call `preventDefault()`
> ... in which case both promises reject with an "AbortError" DOMException, and location.href and navigation.currentEntry stay on their original value.

For the TransitionPrepareEvent, this would abort the navigation.
For the TransitionBeforeSwapEvent, this makes no sense. We could instead set `cancelable = false` for this event, or simply cancel the animation.

For the TransitionBeforeSwapEvent this does not make sense. We could instead set `cancelable = false` for this event ~~or just cancel the animation~~ (we already have skipTransition() in the ViewTransition object if we want to cancel the animation).

Calling `intercept()` will
> Cancel any fragment navigation or cross-document navigation.

The navigation API does many things better than the history API. So we should follow it where it makes sense. However, it has become complicated compared to its origins. It should not be our goal for Astro to mimic it down to the smallest detail. What I see out of scope are the history entries and the intercept options like "commit" or "scroll".

**CHANGE:** That is a strong argument to go with Matthew's suggestion that calling `intercept()` opts out of the default behavior and that `preventDefault()` should not be used for that.  

You may wonder why we have two events with intercept options while the navigation API has only one:
Wouldn't it be possible to get by with a single flow? In the diagram above, we could just move "save scroll position" and "preload style links" to the intercept flow? The argument against this is that the upper flow is outside the ViewTransition and the lower part is inside the ViewTransition. Between these two parts is the red arrow that marks the transition into the ViewTransition, i.e. the ViewTransition is an essential part of Astro's navigation flow. 
What we do with navigate() and the TransitionPrepareEvent is very similar to what the Navigation API provides. 
Our before- and after- swap events are not supported by Chrome's View Transition API. There, you would basically use the promises of the ViewTransition object returned by startViewTransition(). In Astro, ViewTransitions are an integral part of Transitions. In fact, it is rather the other way around: we provide ViewTransitions and therefore need a client-side router as well. The Navigation API does not know about the ViewTransition API. There it is not apparent how multiple navigation interceptors can work together in a ViewTransition. In Astro, the interceptors of the TransitionBeforeSwapEvent allow to push flow steps into the callback of startViewTransition.

**Comment by martrapp** [comment]
In the last few days I have completed a large part of a proof of concept in which `astro-transition-prepare` is the `NavigateEvent` of the navigation API. The PoC shows that this would be a solution with many advantages. However, it also shows that the effort required to simulate the Navigation API on browsers that do not natively support it is far too complicated, even for a subset. To my knowledge, there is no reliable polyfill. 

The benefit of integrating with the navigation API is that in the long run we can get rid of a lot of the code that is currently maintained by Astro but will become redundant by using the API 
* Event handlers for link navigation, history traversal and form submission, 
* Scrolling to the target hash 
  (which the browser can do much better than we can, as it also takes into account that the page may not be complete and tries several times asynchronously), 
* Setting history entries, redirects and rollbacks
* An error model with abort signals and error and success events (not yet part of Astro)
* Read access to older history entries (not yet part of Astro)

Thus, using the `NavigatEvent` for preparation is still the way to go, in my opinion, since it is a web standard. We should do this as soon as the Navigation API is generally available. A mixed-mode implementation of view transition and simulated navigation API does not seem maintainable in the meantime.

**Reply by matthewp** [reply]
Yeah agreed, I wasn't proposing we move to it now.

**Reply by martrapp** [reply]
Yes, me neither. But on the way I had high hopes that only small parts of Navigation API really have to be polyfilled to get a decent integration. That was not true :)

**Comment by martrapp** [comment]
This proposal now defines three new events:
* `astro:before-swap`
* `astro:after-preparation`
* `astro:before-preparation`
They are called at the start of the navigation (`astro:before-preparation`), directly before the startViewTransition (`astro:after-preparation`) and within the View Transition directly at the start (`astro:before-swap`) to complement the existing `astro:after-swap` event.

The astro:after-* events are simple signals, the astro:before-* events enable the modification of parts of the astro view transition implementation (via `loader` and `swap`).

The proposal for the `astro:before-swap` event is now:
* `readonly` `from: URL` The page where the navigation started (renamed to avoid confusion with the navigation API)
* `read-write` `to: URL` The destination of the navigation (renamed to minimize confusion with the navigation API)
* **`readonly`** `direction: Direction | string` the values directly supported by Astro are `'forward'` and `'backward'` but this can be extended to other values. *(edited to 'readonly' at 11/10)*
* `readonly` `navigationType: NavigationTypeString` 
* **`readonly`** **`sourceElement: Element`** **If triggered by a link navigation, the anchor element. If triggered by form submission, the submitter (and if submitter is null, the form element). Can also be set via the `sourceElement` property of the `options` parameter on a call to `navigate()`** *(added 11/11)*
* `readonly` `info: any` If the transition was initiated by a call to `navigate()`, the value of `options.info`. ~~Otherwise `undefined`~~ <ins>Set to an empty object if undefined. Reflects the state of 
`info` object from the end of the `astro:before-preparation` phase</ins> *(edited 11/10)*
* `read-write` `newDocument: Document` The DOM to be transitioned to. Could be an empty DOM if `swap()` manipulates the current DOM in place.
* `read-write` `swap: ()=>void` Initially the standard implementation of the `swap()` operation. 
The task of the swap operation is to update the current DOM, typically to reflect the contents of newDocument. Event listeners can set `swap` to their own implementation. It is the responsibility of this implementation to call the value that was previously stored in this property. In this way, a decorator / chain of resposibility can be built with parts that run before or after the original `swap()`, or even instead of the original `swap()` if the implementation deliberately chooses not to call the earlier `swap()`. 
* `readonly` `viewTransition: ` The [ViewTransition](https://developer.mozilla.org/en-US/docs/Web/API/ViewTransition) object returned by `startViewTransition()`.  

The event is triggered on `window.document` right at the beginning of the view transition. The listeners should only execute short sequences of synchronous code, as EventListeners cannot be `awaited`. Also `swap()` should only execute short sequences of code as the screen freezes during `swap()`. 

Two additional events are proposed for preparing the transition: `astro:before-prepare`and `astro:after-prepare` These events will also be supported once an implementation of Astro's View Transitions build on Navigation API is available. 

The proposal for the `astro:before-preparation` event is as follows:
* `redonly` `from: URL` The page where the navigation started 
* `readwrite` `to: URL` The destination of the navigation. 
* `readwrite` `direction: Direction | string` The values directly supported by Astro are `'forward'` and `'backward'`, but this can be extended to other values. This property is writable.
* `readonly` `navigationType: NavigationTypeString` 
* **`readonly`** **`sourceElement: Element`** **If triggered by a link navigation, the anchor element. If triggered by form submission, the submitter (and if submitter is null, the form element). Can also be set via the `sourceElement` property of the `options` parameter on a call to `navigate()`** *(added 11/11)*
* `readonly` `info: any` If the transition was initiated by a call to `navigate()`, the value of `options.info`. ~~Otherwise undefined~~ <ins>Set to an empty object if undefined.</ins> *(edited 11/10)*
* `read-write` `newDocument:Document` The DOM to be transitioned to. This can be an empty DOM if `swap()` manipulates the current DOM in place.
* **`readonly`** **`formData: FormData`** **Automatically filled in if the navigation was triggered by a form. If the navigation was triggered by a call to `navigate`, the value of `options.formData`.**  *(added at 11/9)*
* `read-write` `loader: ()=>Promise<void>` Initially holds the built-in implementation of the `loader()` operation. The task of the loader is to define a value for `event.newDocument`. Event listeners can set this to their own implementation. It is the task of this implementation to call and await the value that was previously stored in this property. In this way, a Decorator / Chain of Resposibility can be built with parts that run before or after the original `loader()`, or even instead if the implementation deliberately decides not to call the former `loader()`, for example to replace loading from the server with programmatic DOM manipulation. Another example of a decorator is using the original loader and adding image prefetching based on the content of the newDocument. In the case of redirects while loading, the value of the writable variable `event.to` should also be updated to reflect the redirect.

The event is fired on `window.document` right at the beginning of the navigation. The listeners should only execute short sequences of synchronous code, as EventListeners cannot be 'waited for'.  Arbitrary asynchronous code can be executed within `loader()` by defining a decorator callback.

The `astro:after-preparation` imitates the `astro:after-swap` event. It is a standard event without additional attributes or functions. It is triggered at `window.document` just before the view transition begins. The listeners should only execute short sequences of synchronous code, as EventListeners cannot be waited for. This event is ideal, for example, to remove a waiting animation before the view transition API takes a screenshot.

***Added 11/10:***
The `read-write` properties allow listeners to actively influence the navigation process:  
`to`: determines what the browser's address bar should display after navigation (and what should be loaded on reload or history navigation to the target page)
`direction`: a modifier for the animation to be used for the view transition
`newDocument`: what should be rendered on the target page
`loader`: how `newDocument` is generated
`swap`: how the DOM should be updated during the view transition

The initial values of the properties for the `astro:before-swap` event are taken from the final state of the corresponding `astro:before-preparation` event. In particular, changes to the `info` object are not only visible to other `astro:before-preparation` listeners, but also to the `astro:before-swap` listeners. You can use this to communicate between your `astro:before-*` listeners on a per event basis.

**Reply by matthewp** [reply]
I noticed that the `navigate` event doesn't have a `from`, I suppose that's because the URL is not swapped at that point, so you know the `from` because it's the `location`. In which case, do we need `from` in either of these events?

That's a bit of a nit, otherwise these sound nice!

**Reply by martrapp** [reply]
Yes, `from` is usually the same as `location`. But with history traversals, `popstate` is triggered after `window.location` is set to the new location. Now `location` and `to` are identical and you wouldn't know where you came from.

Navigation API also provides a `from`. It is not part of the event, but can be found in the temporary transition object that only exists during navigation: `window.navigation.transition.from.url`

**Comment by martrapp** [comment]
I just added `formData` to the `astro:before-preparation` event

**Comment by martrapp** [comment]
I just marked `direction` as `readonly` in the `astro:before-swap` event. While updating the value would work for  built-in view transitions, the change comes too late for the fallback solution. There the "old" animation starts before the DOM is changed. You can still use the `astro:before-preparation` event to write to `direction` during navigation processing.  

Plus: Clarification on how property values flow from `astro:before-preparation` event to the according `astro:before-swap` event.

**Comment by fprl** [comment]
Hey everyone,

I've been following the discussion about the new events you are going to add (`astro:before-swap`, `astro:after-preparation`, and `astro:before-preparation`). I was going to comment about a few events that were almost the same:
```
astro:page-start
astro:page-error
astro:page-end
astro:form-start
astro:form-error
astro:form-end
```

I was wondering if you think its a good idea to split the events between pages transition and forms submissions (maybe `navigationType` do the trick). I can imagine this could help to build things like `useRouter` and `useForm` for all UI frameworks inside Astro. For instance, consider two simple examples that leverages these proposed events; one framework agnostic and the other one in React:

### Framework agnostic:
Using nanostores like `createSearchParams` in nanostores/router: https://github.com/nanostores/router/blob/83699283a01f319e080a97fb1f2eef6801af5349/index.js#L162C56-L162C56

### React
```typescript
import { useState, useEffect } from 'react';

function useForm(formId) {
    const [formStatus, setFormStatus] = useState({
        from: null,
        to: null,
        direction: '',
        navigationType: '',
        info: {},
        newDocument: null,
        formData: null,
        error: null,
        loading: false,
        data: null
    });

    useEffect(() => {
        function handleFormStart(event) {
            if (event.detail.formId === formId) {
                setFormStatus(prevStatus => ({
                    ...prevStatus,
                    from: event.detail.from,
                    to: event.detail.to,
                    direction: event.detail.direction,
                    navigationType: event.detail.navigationType,
                    info: event.detail.info || {},
                    formData: event.detail.formData,
                    loading: true,
                    error: null
                }));
            }
        }

        function handleFormError(event) {
            if (event.detail.formId === formId) {
                setFormStatus(prevStatus => ({
                    ...prevStatus,
                    error: event.detail.error,
                    loading: false
                }));
            }
        }

        function handleFormEnd(event) {
            if (event.detail.formId === formId) {
                setFormStatus(prevStatus => ({
                    ...prevStatus,
                    loading: false,
                    data: event.detail.data
                }));
            }
        }

        window.addEventListener('astro:form-start', handleFormStart);
        window.addEventListener('astro:form-error', handleFormError);
        window.addEventListener('astro:form-end', handleFormEnd);

        return () => {
            window.removeEventListener('astro:form-start', handleFormStart);
            window.removeEventListener('astro:form-error', handleFormError);
            window.removeEventListener('astro:form-end', handleFormEnd);
        };
    }, [formId]);

    return { formStatus };
}

export default useForm;
```

PS: In the context of [ViewTransitions form handling](https://github.com/withastro/astro/pull/8963), how could Astro handle endpoint responses in non-HTML formats, such as JSON or other media types? I know you can opt out of router transition with `data-astro-reload`, but then it won't work without JS. I think this comment may be related, if you have a hook that gets the data and you can do things with JS, you should be able to do it without JS (PE).

PS2: Maybe I'm confused and mixing things, sorry for that :)

Thank you!

**Reply by martrapp** [reply]
Hello @fprl, thank you for this contribution to the discussion! Indeed, the view transitions are by default only intended for cases where the target is an HTML document. The events should give custom code more control over this default behavior, so you might be able to build your view on top of that. Better than `navigationType`, the newly proposed `sourceElement` property could help to distinguish form submission from link navigation.

**Reply by matthewp** [reply]
@fprl returning non HTML from a form won't work without JavaScript anyways. If you return JSON I thin k the browser will download the contents (not sure), but it's never what you want.

**Reply by fprl** [reply]
@matthewp I'm aware of that but this doesn't mean its not possible, of course you need some help from the framework:
- https://qwik.builder.io/docs/action/
- https://remix.run/docs/en/main/hooks/use-action-data

The other day I was testing it and it was possible with a `useForm` hook or `Form` component that intercepts submissions + the help of middleware and locals to store the form state in case JS is disabled or the page is not hydrated.

**Comment by martrapp** [comment]
Added `sourceElement` to the proposal as in https://chromium.googlesource.com/chromium/src/+/3719b4f51e4919413e8eb08dab383c5278c756a1

**Reply by matthewp** [reply]
I was hoping for something like this when experimenting with `navigate`! Glad to see it exists now.

**Comment by florian-lefebvre** [comment]
@martrapp can we close this proposal? I think we have all those events now

**Reply by martrapp** [reply]
Hi Florian, yes we can! 😀

**Reply by florian-lefebvre** [reply]
Awesome thanks for confirming!
